### ISSUE TITLE
AOT Autograd partitioned forward pass outputs same tensor twice

### ISSUE THREAD

**Issue by anijain2305** [pytorch/functorch#421]
~~~
def f(scores, mask):
    scores = scores.masked_fill(mask == 0, 1e-9)
    return torch.nn.functional.softmax(scores, dim=-1)
~~~

~~~


def forward(self, primals_1, primals_2):
    eq = torch.ops.aten.eq(primals_2, 0);  primals_2 = None
    clone = torch.ops.aten.clone(primals_1, memory_format = 0);  primals_1 = None
    masked_fill_ = torch.ops.aten.masked_fill_(clone, eq, 1e-09);  clone = None
    _softmax = torch.ops.aten._softmax(masked_fill_, -1, False);  masked_fill_ = None
    return [_softmax, _softmax, eq]

~~~

As seen above, _softmax output is duplicated twice. This results in unnecessary memory transfers. This happens when the output of the forward pass also needs to be saved for the backward pass. Partitioning algo does not account for this corner case.

**Comment by Chillee** [comment]
well, this doesn't result in unnecessary memory transfer, right? Just some extra (minor) overhead of putting a tensor in a list twice.

**Comment by anijain2305** [comment]
Might be. I suspected it because of following results

![image](https://user-images.githubusercontent.com/13822661/151471144-41c6f8c0-df3a-4796-8436-031039f1ff0c.png)


From eyeballing the graphs, there is no fusion anywhere. So, it should have been as fast as baseline.

**Comment by Chillee** [comment]
Could be overhead? Or perhaps an inefficient partitioning...
